File: mailblastr-cli/src/commands/segments.js
'use strict';

const { parseJson, clean, withPagination, pagination } = require('../helpers');

function register({ group, leaf, act }) {
  const segments = group('segments', 'Manage contact segments (scoped to a sending domain)');

  act(
    leaf(segments, 'create', 'Create a segment on a sending domain')
      .requiredOption('--domain <domain>', 'the sending domain this segment belongs to')
      .requiredOption('--name <name>', 'segment name (unique within the domain)')
      .option('--filter <json>', "filter as JSON, e.g. '{\"status\":\"subscribed\"}'"),
    ({ client, opts }) =>
      client.segments.create(
        clean({ domain: opts.domain, name: opts.name, filter: parseJson(opts.filter, '--filter') }),
      ),
  );

  act(
    withPagination(
      leaf(segments, 'list', "List a domain's segments").requiredOption(
        '--domain <domain>',
        'the sending domain whose segments to list',
      ),
    ),
    ({ client, opts }) => client.segments.list({ domain: opts.domain, ...pagination(opts) }),
  );

  act(leaf(segments, 'get <id>', 'Retrieve a segment'), ({ client, args: [id] }) =>
    client.segments.get(id),
  );

  act(
    leaf(segments, 'contacts <id>', 'Preview the contacts a segment currently resolves to'),
    ({ client, args: [id] }) => client.segments.contacts(id),
  );

  act(
    leaf(segments, 'update <id>', 'Update a segment')
      .option('--name <name>', 'new name')
      .option('--filter <json>', 'new filter as JSON'),
    ({ client, opts, args: [id] }) =>
      client.segments.update(id, clean({ name: opts.name, filter: parseJson(opts.filter, '--filter') })),
  );

  act(leaf(segments, 'delete <id>', 'Delete a segment'), ({ client, args: [id] }) =>
    client.segments.remove(id),
  );
}

module.exports = { register };

File: mailblastr-ruby/lib/mailblastr/segments.rb
# frozen_string_literal: true

module Mailblastr
  # Segments are DOMAIN-FIRST: `domain` is required on create and list. Names
  # are unique within a domain (reusable across domains); every domain also
  # carries an auto-created "General" (all contacts) segment.
  module Segments
    class << self
      # POST /segments
      #   Mailblastr::Segments.create({ domain: "yourdomain.com", name: "VIP",
      #                                 filter: { status: "subscribed" } })
      def create(params)
        Client.require_domain!(params, "Segments.create")
        Client.request(:post, "/segments", body: params)
      end

      # GET /segments/:id
      def get(segment_id)
        Client.request(:get, "/segments/#{Client.path_escape(segment_id)}")
      end

      # List a domain's segments (`domain` required). GET /segments?domain=
      def list(params)
        domain = Client.require_domain!(params, "Segments.list")
        Client.request(:get, "/segments", query: { domain: domain }.merge(Client.pagination(params)))
      end

      # Preview the contacts a segment currently resolves to (filter matches
      # plus explicit memberships). With no pagination params the response is
      # capped at 1,000 contacts and sets `has_more` — a segment can hold far
      # more than that, so page with `limit` + `after` to read all of it.
      # GET /segments/:id/contacts
      def contacts(segment_id, params = {})
        Client.request(
          :get,
          "/segments/#{Client.path_escape(segment_id)}/contacts",
          query: Client.pagination(params)
        )
      end

      # PATCH /segments/:id
      def update(segment_id, params)
        Client.request(:patch, "/segments/#{Client.path_escape(segment_id)}", body: params)
      end

      # DELETE /segments/:id
      def delete(segment_id)
        Client.request(:delete, "/segments/#{Client.path_escape(segment_id)}")
      end
    end
  end
end
